modalc-lib/modal-vector.rkt

```diff
@@ -0,0 +1,152 @@
+                                   [else value])))))))
+
+(module+ test
+  (require ruinit
+           "test-common.rkt")
+  
+  (define/contract simple-vec
+    (modal-vector/c mode:always integer? integer?)
+    (vector 1 2))
+  (define/contract every-other
+    (modal-vector/c* integer? integer? (and/c positive? integer?) #:ref (mode:once-every 2) #:set (mode:first 2))
+    (vector 4 5 6))
+  (define (get n x)
+    (vector-ref n x))
+  (define (set n x v)
+    (vector-set! n x v))
+  
+  (define/contract basic-vec
+    (modal-vector/c* (and/c positive? integer?) integer? integer? integer? zero? #:ref (mode:once-every 3) #:set (mode:first 8))
+    (vector 2 -3 4.5 1 0))
+  (define/contract complex-vec
+    (modal-vector/c* integer? integer? integer? integer? zero? #:set (mode:first 8) #:ref (mode:once-every 3))
+    (vector #f #t 12 #t 2))
+
+  ;; detects that there's no mode
+  #;(define/contract no-mode-error
+    (modal-vector/c* integer? integer?)
+    (vector 1 1))
+  
+  #;(define/contract no-mode-error2
+    (modal-vector/c any/c integer? integer?)
+    (vector 0 0))
+  
+  (test-begin
+    #:name simple-test
+    (test-equal? (vector-ref simple-vec 1) 2)
+    (vector-set! simple-vec 1 0)
+    (test-equal? (vector-ref simple-vec 1) 0)
+    (test-exn exn:fail:contract:blame? (vector-set! simple-vec 0 'bad))
+    (test-equal? (vector-ref simple-vec 0) 1)
+    (test-exn exn:fail:contract:blame? (vector-set! simple-vec 1 #f))
+    (test-equal? (vector-ref simple-vec 1) 0)
+    (vector-set! simple-vec 0 -10)
+    (vector-set! simple-vec 1 15))
+  
+  (test-begin
+   #:name every-other-test
+   (test-equal? (vector-ref every-other 0) 4) ; yes
+   (test-equal? (vector-ref every-other 1) 5) ; no
+   (vector-set! every-other 1 -2) ; yes
+   (vector-set! every-other 2 2) ; yes
+   (vector-set! every-other 2 -12) ; no
+   (test-equal? (vector-ref every-other 1) -2) ; yes
+   (test-equal? (vector-ref every-other 2) -12) ; no
+   (test-exn exn:fail:contract:blame? (vector-ref every-other 2)) ; yes
+   (vector-set! every-other 2 55) ; no
+   (test-equal? (vector-ref every-other 2) 55) ; yes
+   (vector-set! every-other 0 0))
+  
+  (test-begin
+   #:name get-function
+   (test-equal? (get basic-vec 0) 2) ;; check
+   (test-equal? (get basic-vec 1) -3)
+   (test-equal? (get basic-vec 3) 1)
+   (test-exn exn:fail:contract:blame? (get basic-vec 2)) ;; checks
+   (test-equal? (get basic-vec 2) 4.5) ;; doesn't check
+   (test-equal? (get basic-vec 4) 0)
+   (test-equal? (get complex-vec 2) 12) ;; check
+   (test-equal? (get complex-vec 0) #f)
+   (test-equal? (get complex-vec 4) 2)
+   (test-exn exn:fail:contract:blame? (get complex-vec 4)) ;;check
+   (test-equal? (get complex-vec 1) #t)
+   (test-equal? (get complex-vec 3) #t)
+   (test-exn exn:fail:contract:blame? (get complex-vec 3)))
+
+  ;; (define basic-vec (vector 2 -3 4.5 1 0))
+  ;; (define complex-vec (vector #f #t 12 #t 2))
```

Comment:
dead code?

Reply:
i put that there just to reference what the vectors looked like